FogCreek/Glitch-Community
# Create Feature Flag that is Protected by a Test Team Membership

## Links
* Remix link: https://sponge-credit.glitch.me/
* Issue-tracker link: https://app.clubhouse.io/glitch/story/6208/create-feature-flag
* Specs/doc links: https://docs.google.com/document/d/1a4s0HSkkwq3nuNsX4GqrC7sVrA7AOJ255psYlal9dFc/edit?ts=5e052df8

## Changes:
Hopefully nothing very noticeable :) 
My intention here was to:
- add a new Optimizely attribute that we can use to build a new audience for a new feature flag of Pufferfish
- protect that flag so that it only works if Optimizely says so AND if we enable it on the /secret page. For our other flags we enable it if optimizely says so OR if it's enabled on the /secret page.
- The general gist here is that we want Folks apart of the Test Team to be able to toggle the new code whenever they like on /secret, but that the general public can't get a preview of our code right now. 

## How To Test:
* ~I'm not sure :( I don't see anything different, but my guess is we have to deploy first to see the changes in Optimizely. Maybe @GregWeil you have thoughts, I know you have a lot expertise in this area? Maybe I can run this code in a prod env or something to get Optimizely to pick up the changes?~
* Add this key to your .env file (https://github.com/FogCreek/Glitch-Community/pull/1151#pullrequestreview-339493595) or use the remix where I'll add it myself
* Check out this table, does it mesh with your experience?

| Case  | Expected Outcome | 
| ------------- | ------------- |
| In Testing Team  + haven't done anything on /secret   |  /settings does NOT say pufferfish |
| In Testing Team + Enabled the feature on /secret  | /settings says pufferfish (weird flash of 404 ok for now, getting fixed in a future pr) |
| In Testing Team + Disabled the feature on /secret  | settings does NOT say pufferfish |
| NOT In Testing Team + haven't done anything on /secret   |  /settings does NOT say pufferfish |
| NOT In Testing Team + Enabled the feature on /secret  | settings does NOT say pufferfish  |
| NOT In Testing Team + Disabled the feature on /secret  | settings does NOT say pufferfish |

## Feedback I'm looking for:
- Is there anyway to get the pufferfish feature flag without being a member of the glitch team (through Optimizely cookies or something?). I'm not sure if I need another manual check here in our code. 
- This feels like "A Lot" but my hope here is that if we ever need this additional protection in the future we can reuse this code, does that seem to be the case?
- I copied over a lot of stuff based on other Optimizely features and SSR and loosely followed where all these things were being used, but if I forgot something (or added something I don't need to in this case) feel free to shout!
- Does the /secret page make sense? I feel like we could simplify but it's starting to hurt my brain a bit so maybe we could do that in a future pr or something
- I moved the check for if a user is in the testing team to constants because if I left it in rollouts.js it created a circular dependency with current user, but maybe there's a better place? feel free to let me know if you think of somewhere else it should be

## To do before deploy
I could remove the changes made to /settings before deploying (although I don't think it's a big deal if we keep it?) just wanted to have it so folks could test with it more easily
